pgxn/neon/pagestore_smgr.c

```diff
@@ -1150,18 +1153,23 @@ neon_read_at_lsn(RelFileNode rnode, ForkNumber forkNum, BlockNumber blkno,
 		{
 			/* Combine all prefetch requests with primary request */
 			page_server->send((NeonRequest *) & request);
+			prefetch_lsn = request_lsn;
 			for (i = 0; i < n_prefetch_requests; i++)
 			{
 				request.rnode = prefetch_requests[i].rnode;
 				request.forknum = prefetch_requests[i].forkNum;
 				request.blkno = prefetch_requests[i].blockNum;
+				if (request_latest)
+				{
+					request.req.lsn = GetLastWrittenLSN(request.rnode, request.forknum, request.blkno);
+					prefetch_lsn = Max(prefetch_lsn, request.req.lsn);
+				}
```

Comment:
I don't understand this change. Is this a bugfix? Did we previously use a too-low request LSN, so that we could incorrectly get old page version?

Reply:
Now it is not a bug fix. It is just prefetch improvement.
Assume that there are two pages P1 and P2  with LSNs 100 and 200. Then them are swapped out and in our last written LSN cache we have last_written(P1) = 100, last_written(P2) = 200.
Then we want to load P1 and prefetch P2.
Before this change we request both pages with LSN=100 and set `prefetch_lsn` to 100. So pageserver will not wait until it receives WAL till LSN=200 and returns old image of P2. So this prefetch response will be rejected.
Now we request P2 with LSN=200 and also update `prefetch_lsn` to 200.

Why do we ever need `prefetch_lsn`? It will not be needed if we have infinite last writen lsn cache. But it is not true. We keep only information of most recently updated pages. If we do not find page in last written LSN cache, then we have to use global last written LSN which can be larger than actual LSN of the requested page. To reduce probability of rejecting prefetch page we use compare requested LSN with `Max(request_lsn, GetPageLSN(page))`.

But thank you very much for this notice. Looking once again at this code I understand that  it is not correct to calculate `request_lsn` as maximum of all prefetch request LSNs. We can do it only if we use this max(LSN) for all prefetch requests.
I will submit new fix shortly.